articles/20230820_dev-container-vscode.md

```diff
@@ -0,0 +1,113 @@
+
+Some of the main benefits of using dev containers include:
+
+- **Pre-configured Build Environments**: Dev containers come with a base image that has all the software, tools, and dependencies pre-installed, so you can get started coding right away.
+- **Isolated Environments**: Each dev container has its own isolated filesystem, networking, memory, and CPU, so there are no conflicts with other projects or software on your local machine.
+- **Reproducible Builds**: Dev containers provide the exact same environment every time they're launched, ensuring consistent build results. No more "it works on my machine!" issues.
+- **Less Setup Time**: Starting a new project with dev containers means you can skip the lengthy setup and configuration process. Just open your project in the container and everything is ready to go.
+- **Flexibility**: You can choose a base image with the software and tools you want or build your own custom base image, meeting your specific needs.
+
+Dev containers revolutionize the developer experience by providing pre-configured, isolated environments that can supercharge your productivity. If you haven't tried them yet, you owe it to yourself as a developer to give dev containers a shot. They may just change the way you work! The future is containerized!
+
+## Prerequisites
+
+Before you begin, ensure you have the following:
+
+- **Docker**: Installed and running on your machine.
+- **VSCode**: Installed with the Remote - Containers extension.
+- **Basic Familiarity**: With Docker and VSCode.
+
+## Step 1: Create a `Dockerfile`
+
+A Dockerfile specifies the instructions to create the image for your development environment. Follow these steps:
+
+1. **Create a Directory**: For your project and navigate into it.
+
+   ```sh
+   mkdir my-dev-container && cd my-dev-container
+
+   ```
+
+2. **Create a Dockerfile**: Inside the directory, create a file named Dockerfile with the following content:
+
+   ```sh
+   FROM ubuntu:20.04
+
+   # Avoid warnings by switching to noninteractive
+   ENV DEBIAN_FRONTEND=noninteractive
+
+   # Use the default user 'developer'
+   ARG USERNAME=developer
+   ARG USER_UID=1001
+   ARG USER_GID=$USER_UID
+
+   # Create the user 'developer' with sudo access
+   RUN groupadd --gid $USER_GID $USERNAME \
+       && useradd --uid $USER_UID --gid $USER_GID -m $USERNAME \
+       && apt-get update \
+       && apt-get install -y sudo \
+       && echo $USERNAME ALL=\(root\) NOPASSWD:ALL > /etc/sudoers.d/$USERNAME \
+       && chmod 0440 /etc/sudoers.d/$USERNAME
+
+   # Install development tools
+   RUN apt-get install -y git build-essential
+
+   # Switch back to the dialog frontend for any additional apt-get installations
+   ENV DEBIAN_FRONTEND=dialog
+   ```
+
+## Step 1: Create a `devcontainer.json` File
```

Comment:
create or find one great devcontainer sample (you should reference the source) that works 100%